lib/parallel_test_base.pm

```diff
@@ -0,0 +1,1961 @@
+        my $_temp = 1;
+        $_temp = script_retry("nc -zvD $_guest 22", option => '--kill-after=3 --signal=9', delay => 1, retry => $args{retry}, die => 0);
+        save_screenshot;
+        $_ret |= $_temp;
+        record_info("Guest $_guest ssh failed", "Failed to detect open port 22 on guest $_guest", result => 'fail') if ($_temp != 0);
+        $_temp = script_retry("ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root\@$_guest echo MIGRATION > /tmp/guestok && rm -f -r /tmp/guestok", option => '--kill-after=3 --signal=9', delay => 1, retry => 3, die => 0);
+        save_screenshot;
+        record_info("Guest $_guest can not be accessed", "Failed to open and modify file on guest $_guest", result => 'fail') if ($_temp != 0);
+        $_ret |= $_temp;
+    }
+    croak("ssh connection failed for certain guest") if ($_ret != 0 and $args{die} == 1);
+    return $_ret;
+}
+
+=head2 remove_guest
+
+Remove guest forcibly. Main arguments are guest to be removed, and whether die
+(1) or not (0) if any failures happen. This subroutine also calls construct_uri
+to determine the desired URI to be connected if the interested party is not
+localhost. Please refer to subroutine construct_uri for the arguments related.
+=cut
+
+sub remove_guest {
+    my ($self, %args) = @_;
+    $args{guest} //= '';
+    $args{die} //= 0;
+    $args{driver} //= '';
+    $args{transport} //= 'ssh';
+    $args{user} //= '';
+    $args{host} //= 'localhost';
+    $args{port} //= '';
+    $args{path} //= 'system';
+    $args{extra} //= '';
+    croak("Guest to be removed must be given") if (!$args{guest});
+
+    my $_ret = 0;
+    my $uri = "--connect=" . $self->construct_uri(driver => $args{driver}, transport => $args{transport}, user => $args{user}, host => $args{host}, port => $args{port}, path => $args{path}, extra => $args{extra});
+    foreach my $_guest (split(/ /, $args{guest})) {
+        my $_temp = 1;
+        script_run("virsh $uri destroy $_guest");
+        $_temp = script_run("virsh $uri list --all | grep $_guest") == 0 ? script_run("virsh $uri undefine $_guest || virsh $uri undefine $_guest --keep-nvram") : 0;
+        $_temp |= script_run("xl -vvv destroy $_guest") if (script_run("xl list | grep $_guest") == 0);
+        $_ret |= $_temp;
+        save_screenshot;
+        record_info("Guest $_guest removing failed", "Failed to remove guest $_guest", result => 'fail') if ($_temp != 0);
+    }
+    croak("Failed to remove all guests") if ($_ret != 0 and $args{die} == 1);
+    return $_ret;
+}
+
+=head2 shutdown_guest
+
+Shutdown guest and verify result. Main arguments are guest to be powered off,
+virtualization management tool (virsh or xl) to be used and whether die (1) or
+not (0) if any failures happen. This subroutine also calls construct_uri to
+determine the desired URI to be connected if the interested party is not
+localhost. Please refer to subroutine construct_uri for the arguments related.
+=cut
+
+sub shutdown_guest {
```

Comment:
For these  general apis in this block, create_guest/define_guest/shutdown_guest/etc, before implementing them in this new module parallel_test_base.pm which serves migration/parallel mainly, is it being considered extending the existing library functions of similar functionality , or extending the existing library files by putting them there so that non-parallel jobs can use them too?

Reply:
Extending existing functions might not be good enough, because those existing ones are too shabby. I will consider putting these new ones in library files if feasible (Needs fulfill test purpose for guest migration in the first place).